stdlib/public/Concurrency/AsyncStreamBuffer.swift

```
@@ -236,22 +235,20 @@ extension AsyncStream {
 
     func next(_ continuation: UnsafeContinuation<Element?, Never>) {
       lock()
-      if state.continuation == nil {
-        if state.pending.count > 0 {
-          let toSend = state.pending.removeFirst()
-          unlock()
-          continuation.resume(returning: toSend)
-        } else if state.terminal {
-          unlock()
-          continuation.resume(returning: nil)
-        } else {
-          state.continuation = continuation
-          unlock()
-        }
+      state.continuations.append(continuation)
+      if state.pending.count > 0 {
+        let cont = state.continuations.removeFirst()
+        let toSend = state.pending.removeFirst()
+        unlock()
+        cont.resume(returning: toSend)
+      } else if state.terminal {
+        let cont = state.continuations.removeFirst()
+        unlock()
+        cont.resume(returning: nil)
       } else {
         unlock()
-        fatalError("attempt to await next() on more than one task")
```

@phausler apologies for resurrecting this, but if you have a chance, could you help me understand why this check was removed? i assume it was not directly related to the conditional Sendable conformance since the same change wasn't made to the throwing variant, but maybe i'm overlooking something.

---

I'd like the `fatalError` back please.

---

I have a contrary opinion on this.

I would like this same behavior/changeset to be applied to the `AsyncThrowingStream._Storage`.

So far as I can reason, there is no reason that consuming a stream from multiple tasks should be an error. The current behavior of AsyncStream in that scenario is to distribute the stream elements to each consumer somewhat randomly - which I think is perfectly reasonable behavior.

I don't think fatalError in this circumstance is necessary or beneficial.

I'm sympathetic to the argument that multi-consumption is a violation of this API and should fail fast to alert the user to a possible bug. But I also believe that multi-consumption _ought_ to be a feature - so if it isn't this API then it deserves to be first-classed in some other way.

Removing a fatalError here seems like a relatively low-risk (?) way to make progress on multi-consumption.

At any rate, AsyncStream and AsyncThrowing stream should definitely have the same behavior in a multi-consumer scenario - and right now they do not.

---

Looks like this is outstanding: https://github.com/swiftlang/swift/pull/76571

---

Multi consumption means shared state and the tool for that is actor.

---

I'm not sure I understand what that has to do with the `fatalError` here?

---

Ah - I see. Yeah I was not thinking about that and I agree.
In other words, in order to multicast a stream of `Element`, `Element` must be `Sendable` (or perhaps "Cloneable" if the performance hit is acceptable).